File: source/android/app/src/main/java/com/qppd/ricedryer/ui/auth/AuthViewModel.java
package com.qppd.ricedryer.ui.auth;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.qppd.ricedryer.data.repository.AuthRepository;

public class AuthViewModel extends ViewModel {
    private final AuthRepository authRepository;
    private final MutableLiveData<AuthResult> loginResult;
    private final MutableLiveData<AuthResult> registerResult;
    private final MutableLiveData<AuthResult> resetPasswordResult;

    public AuthViewModel() {
        authRepository = AuthRepository.getInstance();
        loginResult = new MutableLiveData<>();
        registerResult = new MutableLiveData<>();
        resetPasswordResult = new MutableLiveData<>();
    }

    public LiveData<AuthResult> getLoginResult() {
        return loginResult;
    }

    public LiveData<AuthResult> getRegisterResult() {
        return registerResult;
    }

    public LiveData<AuthResult> getResetPasswordResult() {
        return resetPasswordResult;
    }

    public void login(String email, String password) {
        authRepository.login(email, password)
                .addOnSuccessListener(authResult -> {
                    loginResult.setValue(new AuthResult(true, null));
                })
                .addOnFailureListener(e -> {
                    loginResult.setValue(new AuthResult(false, e.getMessage()));
                });
    }

    public void register(String email, String password, String name) {
        authRepository.register(email, password, name)
                .addOnSuccessListener(authResult -> {
                    registerResult.setValue(new AuthResult(true, null));
                })
                .addOnFailureListener(e -> {
                    registerResult.setValue(new AuthResult(false, e.getMessage()));
                });
    }

    public void resetPassword(String email) {
        authRepository.sendPasswordResetEmail(email)
                .addOnSuccessListener(aVoid -> {
                    resetPasswordResult.setValue(new AuthResult(true, null));
                })
                .addOnFailureListener(e -> {
                    resetPasswordResult.setValue(new AuthResult(false, e.getMessage()));
                });
    }

    public boolean isUserLoggedIn() {
        return authRepository.isUserLoggedIn();
    }

    public static class AuthResult {
        private final boolean success;
        private final String error;

        public AuthResult(boolean success, String error) {
            this.success = success;
            this.error = error;
        }

        public boolean isSuccess() {
            return success;
        }

        public String getError() {
            return error;
        }
    }
}

File: source/android/RiceDryer/app/src/main/java/com/qppd/ricedryer/ui/auth/AuthViewModel.kt
package com.qppd.ricedryer.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.auth.FirebaseUser
import com.qppd.ricedryer.data.repository.AuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed class AuthState {
    object Idle : AuthState()
    object Loading : AuthState()
    data class Success(val user: FirebaseUser) : AuthState()
    data class Error(val message: String) : AuthState()
}

class AuthViewModel(private val authRepository: AuthRepository = AuthRepository()) : ViewModel() {
    
    private val _authState = MutableStateFlow<AuthState>(AuthState.Idle)
    val authState: StateFlow<AuthState> = _authState.asStateFlow()
    
    val currentUser = authRepository.currentUser
    
    fun signUp(email: String, password: String, displayName: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.signUp(email, password, displayName)
            _authState.value = result.fold(
                onSuccess = { AuthState.Success(it) },
                onFailure = { AuthState.Error(it.message ?: "Sign up failed") }
            )
        }
    }
    
    fun signIn(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.signIn(email, password)
            _authState.value = result.fold(
                onSuccess = { AuthState.Success(it) },
                onFailure = { AuthState.Error(it.message ?: "Sign in failed") }
            )
        }
    }
    
    fun sendPasswordResetEmail(email: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            val result = authRepository.sendPasswordResetEmail(email)
            _authState.value = result.fold(
                onSuccess = { AuthState.Success(authRepository.getCurrentUser()!!) },
                onFailure = { AuthState.Error(it.message ?: "Password reset failed") }
            )
        }
    }
    
    fun signOut() {
        authRepository.signOut()
        _authState.value = AuthState.Idle
    }
    
    fun resetAuthState() {
        _authState.value = AuthState.Idle
    }
    
    fun isUserSignedIn(): Boolean = authRepository.isUserSignedIn()
}
